Memoryview (#2931)
Codecademy/docs: content/python/concepts/built-in-functions/terms/memoryview/memoryview.md

---
Title: 'memoryview()'

> Description: 'Creates a 'memory view' object that allows Python code to access the internal data of an object's buffer without making a copy of it'

Reviewer: ```suggestion
Description: 'Creates a memoryview object that allows Python code to access the internal data of an object without making a copy of it.'
```
Single quotes in description causes problems with the metadata.

Subjects:
  - 'Data Types'
  - 'Computer Science'
  - 'Built-in Functions'

> Tags:

Reviewer: You are missing tags in the metadata. Please choose two or three from [this list](https://github.com/Codecademy/docs/blob/main/documentation/tags.md).

> CatalogContent:

Reviewer: ```suggestion
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
```

---

> Creates a `memory view` object that allows Python code to access the internal data of an object's buffer without making a copy of it.

Reviewer: ```suggestion
The **`memoryview()`** built-in function creates a `memoryview` object that allows Python code to access the internal data of an object's buffer without making a copy of it.
```

## Syntax

> ```python

Reviewer: The syntax section should be something more like this:

````
```pseudo
result = memoryview(obj)
```

Where `obj`  is the object to create a `memoryview` of, and `result` is the `memoryview` created.
````

obj = bytes([65, 66, 67, 68, 69])
memoryview(obj)

# output <memory at 0x1112b7e80>
```
## Example 1

> `memory view` can only be used on objects that support the buffer protocol and have a readable or writable buffer interface, such as `byte`, `bytearray`, `array.array`, `numpy` arrays, and other objects created using the struct module. `list` and `str` objects don't support buffer protocol.

Reviewer: ```suggestion
> **Note:** `memoryview()` can only be used on objects that support the buffer protocol and have a readable or writable buffer interface, such as `byte`, `bytearray`, `array.array`, `numpy` arrays, and other objects created using the struct module. `list` and `str` objects don't support the buffer protocol.
```
In addition to this edit, move this line before the "Example" header so it is at the end of the "Syntax" section.

> ```python

Reviewer: After moving the prior line to the syntax section, add a new sentence summarizing what the example is doing.

> l = [1, 2, 3, 4, 5, 6]

Reviewer: This needs to be two separate code blocks because an error will stop the interpreter and it will never get to the second part of the example.

memoryview(l)

> #output TypeError: memoryview: a bytes-like object is required, not 'list'

Reviewer: Instead of putting the output in comments in the code, we'd prefer it to be in a separate ` ```shell` block after the example code. Like this:

````
This code results in the following output:

```shell
Output from the code.
```
````

> import numpy as np

Reviewer: This code should be in a seprate block from the one that throws the error, and the output should be in a separate ` ```shell` block.

arr = np.array([1, 2, 3, 4, 5, 6])
memoryview(arr)

> #output <memory at 0x117003700>

Reviewer: Note that the resultant memory address will be different for anyone that runs this code.

```

> 
> 
> ## Example 2
> 

Reviewer: ```suggestion

## Example 2

```

> `memory view` supports indexing and slicing operations: 

Reviewer: ```suggestion
`memoryview` supports indexing and slicing operations: 
```

> ```python

Reviewer: Like my prior comment, can we put the output in a separate ` ```shell` block after the code.

arr = bytearray(10)
mv = memoryview(arr)

> mv[0]=mv[-1] = 65

Reviewer: ```suggestion
mv[0] = mv[-1] = 65
```

print(arr)
# output bytearray(b'A\x00\x00\x00\x00\x00\x00\x00\x00A')

print(mv[0])
#output 65

print(mv[1:4])

> #output <memory at 0x117003880>

Reviewer: Note that the resultant memory address will be different for anyone that runs this code.

Author: @SSwiniarski I understand that output in these last 2 code snippets will be different from the output from my IDE for everyone who runs it  but unlike other mistakes I don't exactly understand how i should fix this one. Could you please suggest something?

Reviewer: Just note in the description of the output that it will be different.

---
Title: 'memoryview()'
Description: 'Creates a memoryview object that allows Python code to access the internal data of an object without making a copy of it.'
Subjects:
  - 'Data Types'
  - 'Computer Science'
  - 'Built-in Functions'
Tags:
  - 'Data Types'
  - 'Memory'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
---

The **`memoryview()`** built-in function creates a `memoryview` object that allows Python code to access the internal data of an object's buffer without making a copy of it.

## Syntax

> ```python
> obj = bytes([65, 66, 67, 68, 69])
> mv = memoryview(obj)
> print(mv)
> ```
> This code results in the following output:
> ```shell
> output <memory at 0x117003340>
> ```
> Where `obj` is the object to create a `memoryview` of, and `mv` is the `memoryview` created.

Reviewer: ````suggestion
```pseudo
mv = memoryview(obj)
```

Where `obj` is the object to create a `memoryview` of, and `mv` is the `memoryview` created.
````

> **Note:** `memoryview()` can only be used on objects that support the buffer protocol and have a readable or writable buffer interface, such as `byte`, `bytearray`, `array.array`, `numpy` arrays, and other objects created using the struct module. `list` and `str` objects don't support the buffer protocol.

## Example 1

> 

Reviewer: Need an introductory sentence summarizing what the example is doing.

```python
l = [1, 2, 3, 4, 5, 6]
memoryview(l)
```

> The output will be:

Reviewer: ```suggestion

The output will be:

```

```shell
output TypeError: memoryview: a bytes-like object is required, not 'list'
```

> ```python

Reviewer: Need a sentence before this code block summarizing what it is doing.

import numpy as np
arr = np.array([1, 2, 3, 4, 5, 6])
memoryview(arr)
```

> This will print the following output:

Reviewer: ```suggestion

This will print the following output:

```

```shell
output <memory at 0x117003700>
```

> > **Note** that memory address generated will vary for each execution of the code. This discrepancy is related to the creation of a new memoryview object.

Reviewer: ```suggestion

> **Note** that memory address generated will vary for each execution of the code. This discrepancy is related to the creation of a new memoryview object.
```

## Example 2

`memoryview` supports indexing and slicing operations: 

> ```python

Reviewer: Put the output of this code block in a separate ` ```shell` block rather than in code comments, like the prior examples.

## Example 1

The example below illustrates the creation of memoryview objects, their display, and the outcome when an inappropriate object is passed to the function:

```python
l = [1, 2, 3, 4, 5, 6]
memoryview(l)
```

The output will be:

```shell
TypeError: memoryview: a bytes-like object is required, not 'list'
```

The following code snippet demonstrates how a `memoryview` object is created from a numpy array:

```python
import numpy as np
arr = np.array([1, 2, 3, 4, 5, 6])
memoryview(arr)
```

This will print the following output:

```shell
<memory at 0x117003700>
```

> **Note** that memory address generated will vary for each execution of the code. This discrepancy is related to the creation of a new memoryview object.

## Example 2

`memoryview` supports indexing and slicing operations: 

```python
arr = bytearray(10)
mv = memoryview(arr)
mv[0] = 65
print(mv[0])
```

This will output a modified element: 

```shell
65
```

Here is how a slicing operation on a `memoryview` object is performed:

```python
print(mv[1:4])
```

Note that a slice of a `memoryview` object is also a `memoryview` object:

```shell
<memory at 0x117003880>
```

## Summary

> The 'memoryview' class offers large performance gains when operating on large objects since it doesn’t create a copy when slicing. However using it requires good knowledge of the structure and format of the data in the source object.

Reviewer 2: I would suggest moving this up to line 17.
